### PR TITLE
Add basic physics joint component include `FixedJoint`, `SpringJoint`, `HingeJoint` 

### PR DESCRIPTION
**Pull request by yangfengzzz**
### Please check if the PR fulfills these requirements

- [ ] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Other information:

### CODE REVIEW

**Comment by GuoLei1990** [comment]
```ts
/**
 * A base class providing common functionality for joints.
 * @decorator `@dependentComponents(Collider)`
 */
@dependentComponents(Collider)
export class Joint extends Component {
  /**
   * The connected collider.
   */
  get connectedCollider(): Collider {}
  set connectedCollider(value: Collider) {}

  /**
   *  The scale to apply to the inverse mass of collider 0 for resolving this constraint.
   */
  get connectedMassScale(): number {}
  set connectedMassScale(value: number) {}

  /**
   * The scale to apply to the inverse inertia of collider0 for resolving this constraint.
   */
  get connectedInertiaScale(): number {}
  set connectedInertiaScale(value: number) {}

  /**
   * The scale to apply to the inverse mass of collider 1 for resolving this constraint.
   */
  get massScale(): number {}
  set massScale(value: number) {}

  /**
   * The scale to apply to the inverse inertia of collider1 for resolving this constraint.
   */
  get inertiaScale(): number {}
  set inertiaScale(value: number) {}

  /**
   * The maximum force the joint can apply before breaking.
   */
  get breakForce(): number {}
  set breakForce(value: number) {}

  /**
   * The maximum torque the joint can apply before breaking.
   */
  get breakTorque(): number {}
  set breakTorque(value: number) {}
}

/*
 * A fixed joint permits no relative movement between two bodies. ie the bodies are glued together.
 */
export class FixedJoint extends Joint {

}

export class SpringJoint extends Joint {

  /**
   * The connected anchor position.
   * @note If connectedCollider is set, this anchor is relative offset.
   * Or the anchor is world anchor position.
   */
  get connectedAnchor(): Vector3 {}

  set connectedAnchor(value: Vector3) {}

  /**
   * The swing offset.
   */
  get swingOffset(): Vector3 {}

  set swingOffset(value: Vector3) {}

  /**
   * The minimum distance.
   */
  get minDistance(): number {}

  set minDistance(value: number) {}

  /**
   * The maximum distance.
   */
  get maxDistance(): number {}

  set maxDistance(value: number) {}

  /**
   * The distance beyond the allowed range at which the joint becomes active.
   */
  get tolerance(): number {}

  set tolerance(value: number) {}

  /**
   * The spring strength of the joint.
   */
  get stiffness(): number {}

  set stiffness(value: number) {}

  /**
   * The degree of damping of the joint spring of the joint.
   */
  get damping(): number {}

  set damping(value: number) {}
}

/**
 * A joint which behaves in a similar way to a hinge or axle.
 */
 export class HingeJoint extends Joint {


  /**
   * The current angle in degrees of the joint relative to its rest position.
   */
  get angle(): number {}

  /**
   * The angular velocity of the joint in degrees per second.
   */
  get velocity(): ReadOnly<Vector3> {}

  /**
   * The connected anchor position.
   * @note If connectedCollider is set, this anchor is relative offset.
   * Or the anchor is world anchor position.
   */
  get connectedAnchor(): Vector3 {}

  set connectedAnchor(value: Vector3) {}

  /**
   * The anchor rotation.
   */
  get axis(): Vector3 {}

  set axis(value: Vector3) {}

  /**
   * The swing offset.
   */
  get swingOffset(): Vector3 {}

  set swingOffset(value: Vector3) {}

  /**
   * Enables the joint's limits. Disabled by default.
   */
  get useLimits(): boolean {}

  set useLimits(value: boolean) {}

  /**
   * Limit of angular rotation (in degrees) on the hinge joint.
   */
  get limits(): JointLimits {}

  set limits(value: JointLimits) {}

  /**
   * Enables the joint's motor. Disabled by default.
   */
  get useMotor(): boolean {}

  set useMotor(value: boolean) {}

  /**
   * Enables the joint's spring. Disabled by default.
   */
  get useSpring(): boolean {}

  set useSpring(value: boolean) {}

  /**
   * The motor will apply a force up to a maximum force to achieve the target velocity in degrees per second.
   */
  get motor(): JointMotor {}

  set motor(value: JointMotor) {}
}
```

File: packages/physics-physx/src/PhysXPhysics.ts [hunk 1]
```diff
@@ -186,6 +193,48 @@ export class PhysXPhysics {
...
+    position1: Vector3,
+    rotation1: Quaternion
+  ): IFixedJoint {
+    return new PhysXFixedJoint(actor0, position0, rotation0, actor1, position1, rotation1);
+  }
+
+  /**
+   * {@inheritDoc IPhysics.createHingeJoint }
+   */
+  static createHingeJoint(
```
Comment (GuoLei1990) [inline packages/physics-physx/src/PhysXPhysics.ts:213]: Simplified parameters like this:
```ts
createHingeJoint(collider: PhysXCollider, connectedCollider: PhysXCollider)
createSpringJoint(collider: PhysXCollider, connectedCollider: PhysXCollider, connectedAnchor:Vector3, swingOffset:Vector3);
```

File: packages/core/src/physics/enums/ConstraintFlag.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  CollisionEnabled = 8,
+  /** limits for drive strength are forces rather than impulses */
+  DriveLimitsAreForces = 32,
+  /** perform preprocessing for improved accuracy on D6 Slerp Drive (this flag will be removed in a future release when preprocessing is no longer required) */
+  ImprovedSlerp = 64,
+  /** suppress constraint preprocessing, intended for use with rowResponseThreshold. May result in worse solver accuracy for ill-conditioned constraints. */
+  DisablePreprocessing = 128,
+  /** enables extended limit ranges for angular limits (e.g. limit values > PxPi or < -PxPi) */
+  EnableExtendedLimits = 256
+}
```
[the remark this answers is not in the document]
Comment (yangfengzzz) [reply packages/core/src/physics/enums/ConstraintFlag.ts]: done

File: packages/core/src/physics/joint/HingeJoint.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    const swingOffset = this._swingOffset;
+    if (value !== swingOffset) {
+      swingOffset.copyFrom(value);
+    }
+    (<IHingeJoint>this._nativeJoint).setSwingOffset(swingOffset);
+  }
+
+  /**
+   * The connected anchor position.
+   * @note If connectedCollider is set, this anchor is relative offset.
```
[the remark this answers is not in the document]
Comment (yangfengzzz) [reply packages/core/src/physics/joint/HingeJoint.ts]: done

File: packages/core/src/physics/joint/HingeJoint.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+import { dependentComponents } from "../../ComponentsDependencies";
+import { Vector3, Quaternion } from "@oasis-engine/math";
+import { JointMotor } from "./JointMotor";
+import { JointLimits } from "./JointLimits";
+
+/**
+ * A joint which behaves in a similar way to a hinge or axle.
+ * @decorator `@dependentComponents(Collider)`
+ */
+@dependentComponents(Collider)
```
[the remark this answers is not in the document]
Comment (yangfengzzz) [reply packages/core/src/physics/joint/HingeJoint.ts]: done

File: packages/core/src/physics/enums/HingeJointFlag.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Flags specific to the Hinge Joint.
+ */
+export enum HingeJointFlag {
```
Comment (GuoLei1990) [inline packages/core/src/physics/enums/HingeJointFlag.ts:4]: Add @internal, and don't need to export in `index.html`

File: packages/core/src/physics/joint/FixedJoint.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import { Joint } from "./Joint";
+import { IFixedJoint } from "@oasis-engine/design";
+import { PhysicsManager } from "../PhysicsManager";
+import { dependentComponents } from "../../ComponentsDependencies";
+import { Collider } from "../Collider";
+import { Vector3 } from "@oasis-engine/math";
+
+/*
+ * A fixed joint permits no relative movement between two bodies. ie the bodies are glued together.
```
Comment (GuoLei1990) [inline packages/core/src/physics/joint/FixedJoint.ts:9]: comments should: `bodies` -> `colliders`

File: packages/design/src/physics/joints/IHingeJoint.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+   * @param ratio the gear ratio
+   */
+  setDriveGearRatio(ratio: number): void;
+
+  /**
+   * sets a single flag specific to a Revolute Joint.
+   * @param flag The flag to set or clear.
+   * @param value the value to which to set the flag
+   */
+  setRevoluteJointFlag(flag: number, value: boolean): void;
```
[the remark this answers is not in the document]
Comment (yangfengzzz) [reply packages/design/src/physics/joints/IHingeJoint.ts]: done

File: packages/physics-physx/src/PhysXPhysics.ts (hunk 1 above)
Comment (yangfengzzz) [reply packages/physics-physx/src/PhysXPhysics.ts]: done

File: packages/core/src/physics/joint/HingeJoint.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    const swingOffset = this._swingOffset;
+    if (value !== swingOffset) {
+      swingOffset.copyFrom(value);
+    }
+    (<IHingeJoint>this._nativeJoint).setSwingOffset(swingOffset);
+  }
+
+  /**
+   * The connected anchor position.
+   * @remark If connectedCollider is set, this anchor is relative offset.
```
Comment (GuoLei1990) [inline packages/core/src/physics/joint/HingeJoint.ts:54]: @remark->@remarks

File: packages/core/src/physics/joint/FixedJoint.ts (hunk 6 above)
Comment (yangfengzzz) [reply packages/core/src/physics/joint/FixedJoint.ts:9]: done

File: packages/core/src/physics/enums/HingeJointFlag.ts (hunk 5 above)
Comment (yangfengzzz) [reply packages/core/src/physics/enums/HingeJointFlag.ts:4]: done

File: packages/core/src/physics/joint/HingeJoint.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    return this._limits;
+  }
+
+  set limits(value: JointLimits) {
+    this._limits = value;
+    if (this.useSpring) {
+      (<IHingeJoint>this._nativeJoint).setSoftLimit(value.min, value.max, value.stiffness, value.damping);
+    } else {
+      (<IHingeJoint>this._nativeJoint).setHardLimit(value.min, value.max, value.contactDistance);
+    }
```
Comment (GuoLei1990) [inline packages/core/src/physics/joint/HingeJoint.ts:146]: Here has a BUG:
1. set `useSpring  == false`
2. and then set `limits`
3. set `useSpring  == true`

File: packages/core/src/physics/joint/HingeJoint.ts (hunk 8 above)
Comment (yangfengzzz) [reply packages/core/src/physics/joint/HingeJoint.ts]: done

File: packages/core/src/physics/joint/HingeJoint.ts (hunk 9 above)
Comment (yangfengzzz) [reply packages/core/src/physics/joint/HingeJoint.ts:146]: done

File: packages/design/src/physics/joints/IHingeJoint.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
+import { IJoint } from "./IJoint";
+import { Vector3 } from "@oasis-engine/math";
+
+/**
+ * A joint which behaves in a similar way to a hinge or axle.
+ */
+export interface IHingeJoint extends IJoint {
+  /**
+   * The connected anchor position.
+   * @note If connectedCollider is set, this anchor is relative offset.
```
Comment (GuoLei1990) [inline packages/design/src/physics/joints/IHingeJoint.ts:10]: remarks

Comment (yangfengzzz) [reply packages/design/src/physics/joints/IHingeJoint.ts]: done

File: packages/core/src/physics/joint/HingeJoint.ts [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  /**
+   * @override
+   * @internal
+   */
+  _onAwake() {
+    const { _connectedCollider: connectedCollider, _collider: collider } = this;
+    connectedCollider.localPosition = new Vector3();
+    connectedCollider.collider = null;
```
Comment (GuoLei1990) [inline packages/core/src/physics/joint/HingeJoint.ts:159]: Don't need any more!

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**